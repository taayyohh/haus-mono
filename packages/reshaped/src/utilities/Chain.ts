type ID = number;
type Item<T> = {
	previousId?: ID | null;
	nextId?: ID | null;
	data: T;
};

class Chain<T = unknown> {
	chain: Record<ID, Item<T>> = {};

	tailId: ID | null = null;

	idCounter: ID = 0;

	generateId() {
		this.idCounter += 1;
		return this.idCounter;
	}

	getAll() {
		return this.chain;
	}

	get(id: ID) {
		return this.chain[id];
	}

	isLast(id: ID) {
		return this.tailId !== null && id === this.tailId;
	}

	isEmpty() {
		return typeof this.tailId !== "number";
	}

	add(data: T) {
		const previousId = this.tailId;
		const previousItem = previousId && this.get(previousId);
		const id = this.generateId();

		this.chain[id] = { previousId, data };

		if (previousItem) previousItem.nextId = id;
		this.tailId = id;
		return id;
	}

	remove(id: ID) {
		const target = this.chain[id];

		if (!target) return;

		const previousId = target.previousId;
		const previousItem = previousId && this.get(previousId);
		const nextId = target.nextId;
		const nextItem = nextId && this.get(nextId);

		if (previousItem) previousItem.nextId = target.nextId ?? null;
		if (nextItem) nextItem.previousId = target.previousId ?? null;
		if (!nextId) this.tailId = previousId ?? null;

		const data = this.get(id).data;
		delete this.chain[id];

		return data;
	}

	removePreviousTill(id: ID, condition: (item: Item<T>) => boolean): T | undefined {
		const target = this.get(id);
		const data = this.remove(id);

		if (!target || !target.previousId) return data;

		const keepIterating = !condition(target);
		if (keepIterating) return this.removePreviousTill(target.previousId, condition);
		return data;
	}
}

export default Chain;
