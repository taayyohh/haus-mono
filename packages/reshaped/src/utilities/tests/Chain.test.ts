import Chain from "utilities/Chain";

describe("Chain", () => {
	test("gets an item", () => {
		const chain = new Chain();

		const id = chain.add("Test");
		expect(chain.get(id).data).toBe("Test");
	});

	test("validates last item", () => {
		const chain = new Chain();

		const firstId = chain.add("Test");
		const lastId = chain.add("Test 2");

		expect(chain.isLast(firstId)).toBe(false);
		expect(chain.isLast(lastId)).toBe(true);
	});

	test("adds items", () => {
		const chain = new Chain();

		chain.add("Test");
		expect(chain.getAll()).toMatchSnapshot();
	});

	test("removes item in the middle", () => {
		const chain = new Chain();

		chain.add("Test");
		const id = chain.add("Test 2");
		chain.add("Test 3");
		chain.remove(id);

		expect(chain.getAll()).toMatchSnapshot();
	});

	test("removes side items", () => {
		const chain = new Chain();

		const idFirst = chain.add("Test");
		chain.add("Test 2");
		const idLast = chain.add("Test 3");
		chain.remove(idFirst);
		chain.remove(idLast);

		expect(chain.getAll()).toMatchSnapshot();
	});

	test("removes previous elements till condition is true", () => {
		const chain = new Chain();

		chain.add("Test");
		chain.add("Test 2");
		const id = chain.add("Test 3");
		chain.removePreviousTill(id, (item) => item.data === "Test 2");

		expect(chain.getAll()).toMatchSnapshot();
	});
});
