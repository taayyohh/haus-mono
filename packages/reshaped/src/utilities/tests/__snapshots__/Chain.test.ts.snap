// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chain > adds items 1`] = `
{
  "1": {
    "data": "Test",
    "previousId": null,
  },
}
`;

exports[`Chain > removes item in the middle 1`] = `
{
  "1": {
    "data": "Test",
    "nextId": 3,
    "previousId": null,
  },
  "3": {
    "data": "Test 3",
    "previousId": 1,
  },
}
`;

exports[`Chain > removes previous elements till condition is true 1`] = `
{
  "1": {
    "data": "Test",
    "nextId": null,
    "previousId": null,
  },
}
`;

exports[`Chain > removes side items 1`] = `
{
  "2": {
    "data": "Test 2",
    "nextId": null,
    "previousId": null,
  },
}
`;
